File: src/content/3.9/code/scala/snippet08.scala
implicit def storeComonad[S] = new Comonad[Store[S, ?]] {
  def extract[A](wa: Store[S, A]): A = wa match {
    case Store(f, s) => f(s)
  }
  
  def duplicate[A](wa: Store[S, A]): Store[S, Store[S, A]] = wa match {
    case Store(f, s) => Store(Store(f), s)
  }
}

File: src/content/3.9/code/ocaml/snippet08.ml
module Store_comonad (S : sig
  type s
end)
(F : Functor with type 'a t = (S.s, 'a) store) :
  Comonad with type 'a w = (S.s, 'a) store = struct
  type 'a w = (S.s, 'a) store

  include F

  let extract : 'a w -> 'a = fun (Store (f, s)) -> f s

  let duplicate : 'a w -> 'a w w =
   fun (Store (f, s)) -> Store ((fun s -> Store (f, s)), s)
 ;;
end
